File: python3/q_table.py
import pickle


class QTable:
    """Q テーブルです。"""

    @classmethod
    def load(cls, path: str):
        with open(path, mode="rb") as f:
            return pickle.load(f)

    def __init__(self, init_q: float, s_space: int, a_space: int):
        """
        init_q: テーブルの初期値
        s_space: 状態空間の大きさ
        a_space: 行動空間の大きさ
        """
        self.init_q = init_q
        self.s_space = s_space
        self.a_space = a_space

        self.table = [[init_q] * a_space for _ in range(s_space)]

    def save(self, path: str):
        """Q テーブルを path に保存します"""
        with open(path, mode="wb") as f:
            pickle.dump(self, f)

File: rust/src/q_table.rs
use bincode;
use serde::{Deserialize, Serialize};
use std::error;
use std::fs;
use std::fs::File;
use std::io::prelude::*;

#[derive(Serialize, Deserialize)]
pub struct QTable {
    pub s_space: usize,
    pub a_space: usize,
    pub init_q: f64,
    pub table: Vec<Vec<f64>>,
}

impl QTable {
    pub fn new(init_q: f64, s_space: usize, a_space: usize) -> Self {
        QTable {
            s_space,
            a_space,
            init_q,
            table: vec![vec![init_q; a_space]; s_space],
        }
    }

    pub fn load(path: &str) -> Result<Self, Box<dyn error::Error>> {
        let mut file = File::open(path)?;
        let metadata = fs::metadata(path)?;
        let mut data = vec![0; metadata.len() as usize];
        file.read(&mut data)?;

        // HELP!!!
        match bincode::deserialize(&data) {
            Ok(obj) => Ok(obj),
            Err(kind) => Err(kind),
        }
    }

    pub fn save(&self, path: &str) -> Result<(), Box<dyn error::Error>> {
        let data = bincode::serialize(self)?;
        let mut file = File::create(path)?;
        file.write(&data)?;
        file.flush()?;
        Ok(())
    }
}
